yarn-project/ethereum/src/l1_tx_utils/readonly_l1_tx_utils.ts

```diff
@@ -63,6 +63,119 @@ export class ReadOnlyL1TxUtils {
     return this.client.getBlockNumber();
   }
 
+  /**
+   * Analyzes pending transactions and recent fee history to determine a competitive priority fee.
+   * Falls back to network estimate if data is unavailable or fails.
+   * @param networkEstimate - The base network priority fee estimate
+   * @returns A competitive priority fee based on pending txs and recent block history
+   */
+  protected async getCompetitivePriorityFee(networkEstimate: bigint): Promise<bigint> {
+    let competitiveFee = networkEstimate;
+
+    // 1. Check pending transactions (current competition)
+    try {
+      const pendingBlock = await this.client.getBlock({ blockTag: 'pending', includeTransactions: true });
```

Comment:
Let's run both getBlock and getFeeHistory requests in parallel, so we lose as little time as possible in roundtrips, especially during block publishing.